python/ray/experimental/ddp/ddp.py

```diff
@@ -0,0 +1,541 @@
+        ]
+        dag = MultiOutputNode(ends)
+
+    compiled_dag = dag.experimental_compile()
+    it = config.it
+    adag_ddp_weights = []
+    elapses = []
+    for i in range(it):
+        x, y = generate_input_output(config)
+        xs = torch.tensor_split(x, num_actors)
+        ys = torch.tensor_split(y, num_actors)
+        x = time.perf_counter()
+        ref = compiled_dag.execute(*xs, *ys)
+        cur_iter_weights = ray.get(ref)
+        end = time.perf_counter()
+        adag_ddp_weights.append(cur_iter_weights)
+        elapse = end - x
+        elapses.append(elapse)
+
+    print_elapses(elapses, "adag ddp")
+    compiled_dag.teardown()
+
+    for actor in actors:
+        ray.kill(actor)
+    ray.shutdown()
+
+    adag_ddp_weights = get_adag_ddp_weights_per_device(
+        adag_ddp_weights, config.num_actors
+    )
+    return adag_ddp_weights
+
+
+def get_adag_ddp_weights_per_device(
+    weights: List[List[Tuple[torch.Tensor]]], num_actors: int
+):
+    """
+    Check that the model is consistent across all ranks after each iteration
+    of training. Deduplicate the weights after the checks.
+
+    Args:
+        weights: Weights from running aDAG DDP for a number of iterations.
+            `weights[i]` is the weights across all actors for the ith iteration.
+    """
+    weights_per_device = []
+    for single_iter_weights in weights:
+        per_device_len = len(single_iter_weights[0])
+        for i in range(1, num_actors):
+            for j in range(per_device_len):
+                assert torch.equal(
+                    single_iter_weights[0][j], single_iter_weights[i][j]
+                ), f"{single_iter_weights[0][j]} vs. {single_iter_weights[i][j]}"
+        weights_per_device.append(single_iter_weights[0])
+    return weights_per_device
+
+
+def compare_weights(W1, W2, desc, allow_error=False):
+    """
+    Compare the weights after each iteration across different training approaches.
+
+    Args:
+        W1: Weights after each iteration from one approach.
+        W2: Weights after each iteration from the other approach.
+        desc: Description of approaches
+        allow_error: Whether small errors are allowed.
+    """
+    assert len(W1) == len(W2)
+    # w1, w2 are weights after a single iteration for the 1st and 2nd approaches,
+    # respectively.
+    max_diff = 0
+    for w1, w2 in zip(W1, W2):
+        assert len(w1) == len(w2)
+        # t1, t2 are weights of a single layer.
+        for t1, t2 in zip(w1, w2):
+            t1 = t1.to("cpu")
+            t2 = t2.to("cpu")
+            if not allow_error:
+                assert torch.allclose(
+                    t1, t2
+                ), f"{desc} max diff: {torch.max(torch.abs(t1 - t2).flatten())}"
+            elif not torch.allclose(t1, t2):
+                max_diff = max(max_diff, torch.max(torch.abs(t1 - t2).flatten()))
+
+    if max_diff != 0:
+        print(f"{desc} max diff: {max_diff}")
+
+
+def print_elapses(elapses, desc):
```

Comment:
We could print all these numbers into a csv. Then we can have a python script to plot a figure based on these csv data.

Reply:
Printed to CSV. No script to plot yet.